sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ExpressionSet.scala

```diff
@@ -50,7 +50,7 @@ object ExpressionSet {
 class ExpressionSet protected(
     protected val baseSet: mutable.Set[Expression] = new mutable.HashSet,
     protected val originals: mutable.Buffer[Expression] = new ArrayBuffer)
-  extends Set[Expression] {
+  extends Set[Expression] with Serializable {
```

Comment:
If `ExpressionSet` is really serialized only in the case of `LogicalPlan`,  we could move `constraints` from `QueryPlan` to `LogicalPlan` but I'm not sure it's correct way.
Have you ever got any problem because `ExpressionSet` is not `Serializable` ?